### PR TITLE
implemented user-config-utils, and attached it to relevant files

### LINKED ISSUE #46: cli-config-utils class
As a developer, I would like to not waste time inputting configuration every time i use the a program, thus a system that can save such info would be useful.

### PR DESCRIPTION
**Pull request by eetupur**
resolves #46

### CODE REVIEW

**Review by veikkahonkanen — APPROVED** [review]

File: src/classes/os-utils.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import { CommandObj, UserConfigJson, OperatingSystems } from "../interfaces/types";
 import { OsCommands } from "./os-commands";
-import fs from "fs";
+import { userConfigUtils } from "./user-config-utils";
```
Comment (anttileppa) [inline src/classes/os-utils.ts]: Class names should be captitalized

```suggestion
import { UserConfigUtils } from "./user-config-utils";
```

File: src/classes/os-utils.ts [hunk 2]
```diff
@@ -18,7 +18,7 @@ export default class OsUtils {
    */
   static getCommand = async (cmd: string): Promise<string> => {
     try {
-      const userConfig: UserConfigJson = await OsUtils.readUserConfig();
+      const userConfig: UserConfigJson = await userConfigUtils.readUserConfig();
```
Comment (anttileppa) [inline src/classes/os-utils.ts]: Class names should be captitalized
```suggestion
      const userConfig: UserConfigJson = await UserConfigUtils.readUserConfig();
```

File: src/classes/os-utils.ts [hunk 3]
```diff
@@ -33,7 +33,7 @@ export default class OsUtils {
    */
   public static getOS = async () : Promise<string | null> => {
     try {
-      const os : UserConfigJson = await OsUtils.readUserConfig();
+      const os : UserConfigJson = await userConfigUtils.readUserConfig();
```
Comment (anttileppa) [inline src/classes/os-utils.ts]: Class names should be captitalized
```suggestion
      const os : UserConfigJson = await UserConfigUtils.readUserConfig();
```

File: src/classes/os-utils.ts [hunk 4]
```diff
@@ -53,40 +53,19 @@ export default class OsUtils {
...
 
-    /**
+  /**
    * Re-writes user config using previous user config as basis
    * 
    * @param os is the OS that is being switched to, if supported
    */
   private static async swapOs (os : string) {
-    const userConfig : UserConfigJson = await OsUtils.readUserConfig();
+    const userConfig : UserConfigJson = await userConfigUtils.readUserConfig();
```
Comment (anttileppa) [inline src/classes/os-utils.ts]: Class names should be captitalized
```suggestion
    const userConfig : UserConfigJson = await UserConfigUtils.readUserConfig();
```

File: src/classes/os-utils.ts [hunk 5]
```diff
@@ -53,40 +53,19 @@ export default class OsUtils {
...
-    try {
-      fs.writeFile("./user-config.json", data, "utf8", (err) => {
-        if (err) {
-          throw err;
-        }
-      });
-    } catch (err) {
-      throw new Error("Error when attempting to write file:" + err);
-    }
+    await userConfigUtils.writeUserConfig(data);
```
Comment (anttileppa) [inline src/classes/os-utils.ts]: Class names should be captitalized
```suggestion
    await UserConfigUtils.writeUserConfig(data);
```

File: src/classes/user-config-utils.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import fs from "fs";
+import { UserConfigJson } from "../interfaces/types";
+import path from "path";
+
+const userConfigPath = `.${path.sep}user-config.json`;
+const userConfigTemplate = `.${path.sep}resources${path.sep}user-config-template.json`;
+
+export class userConfigUtils {
```
Comment (anttileppa) [inline src/classes/user-config-utils.ts]: Class names should be captitalized
```suggestion
export class UserConfigUtils {
```

Comment (anttileppa) [inline src/classes/user-config-utils.ts]: Add tsconfig

File: src/classes/user-config-utils.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+  /**
+   * Reads user config
+   * 
+   * @returns user config as a JSON object that has an interface
+   */
+    static async readUserConfig() : Promise<UserConfigJson> {
```
Comment (anttileppa) [inline src/classes/user-config-utils.ts]:
```suggestion
  static async readUserConfig() : Promise<UserConfigJson> {
```

**Review by anttileppa — CHANGES_REQUESTED** [review]

**Comment by anttileppa** [comment]
Tests are not passing

**Review by veikkahonkanen — APPROVED** [review]

File: test/user-config-utils.spec.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import { UserConfigUtils } from "../src/classes/user-config-utils";
+import { UserConfigJson } from "../src/interfaces/types";
+
+const mockUserConfig : UserConfigJson = {
+  "osPref": "LINUX"
+}
```
Comment (veikkahonkanen) [inline test/user-config-utils.spec.ts]: remove mockUserConfig

File: test/user-config-utils.spec.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+it("reads user-config.json", async () => {
+  const data : UserConfigJson = await UserConfigUtils.readUserConfig();
+
+  expect(data).toHaveProperty("osPref");
+});
+
+it("is able to write to user-config-json", async () => {
+  UserConfigUtils.writeUserConfig = jest.fn();
```
Comment (veikkahonkanen) [inline test/user-config-utils.spec.ts:11]: Add mockUserConfig to line 15 (not used outside of this function)

File: src/classes/os-utils.ts [hunk 10]
```diff
@@ -1,5 +1,6 @@
 import { UserConfigJson, OperatingSystems } from "../interfaces/types";
 import { OsCommands } from "./os-commands";
+import { UserConfigUtils } from "./user-config-utils";
 import fs from "fs";
```
Comment (veikkahonkanen) [inline src/classes/os-utils.ts]: Reamove fs if not used (Sonar doesn't seem to approve of it either)

**Review by veikkahonkanen — APPROVED** [review]

**Review by veikkahonkanen — APPROVED** [review]

File: src/classes/os-utils.ts [hunk 11]
```diff
@@ -1,6 +1,6 @@
 import { UserConfigJson, OperatingSystems } from "../interfaces/types";
 import { OsCommands } from "./os-commands";
-import fs from "fs";
+import { UserConfigUtils } from "./user-config-utils";
 import { windows, linux, macos } from 'platform-detect';
```
Comment (jarinykanen) [inline src/classes/os-utils.ts]:
```suggestion
import { windows, linux, macos } from "platform-detect";
```

File: src/classes/os-utils.ts [hunk 12]
```diff
@@ -85,34 +85,13 @@ export default class OsUtils {
...
-   */
-  private static async readUserConfig() : Promise<UserConfigJson> {
-    try {
-      const data = fs.readFileSync("./user-config.json", "utf8");
-      return JSON.parse(data.toString());
-    } catch(err) {
-      throw new Error ("User config not found: " + err);
-    }
-  }
+
```
Comment (jarinykanen) [inline src/classes/os-utils.ts]:
```suggestion
```

File: src/classes/os-utils.ts [hunk 13]
```diff
@@ -85,34 +85,13 @@ export default class OsUtils {
...
-      fs.writeFile("./user-config.json", data, "utf8", (err) => {
-        if (err) {
-          throw err;
-        }
-      });
-    } catch (err) {
-      throw new Error("Error when attempting to write file:" + err);
-    }
+    await UserConfigUtils.writeUserConfig(data);
   }
```
Comment (jarinykanen) [inline src/classes/os-utils.ts:93]:
```suggestion
```

File: src/classes/user-config-utils.ts [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+const userConfigTemplate = `.${path.sep}resources${path.sep}user-config-template.json`;
+
+/**
+ * Offers functions to read and write the cli's user config
+ * 
+ * @function readUserConfig runs a check after which it reads and returns the contents of user-config.json
+ * 
+ * @function writeUserConfig takes the edited version of the contents of user-config.json and writes them after a check
+ * 
+ * @function checkUserConfig checks that user-config.json exists and creates it, if it doesn't
```
Comment (jarinykanen) [inline src/classes/user-config-utils.ts]: Move function descriptions out from class docs and to the functions themselves.

File: src/classes/user-config-utils.ts [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+ */
+export class UserConfigUtils {
+
+  /**
+   * Reads user config
+   * 
+   * @returns user config as a JSON object that has an interface
+   */
+  static async readUserConfig() : Promise<UserConfigJson> {
```
Comment (jarinykanen) [inline src/classes/user-config-utils.ts]:
```suggestion
  static async readUserConfig(): Promise<UserConfigJson> {
```

File: src/classes/user-config-utils.ts [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+      throw new Error ("Error while reading user-config: " + err);
+    }
+  }
+
+  /**
+   * Writes to user config
+   * 
+   * @param data should receive the edited version of user-config.json 
+   */
+  static async writeUserConfig(data : string) : Promise<void> {
```
Comment (jarinykanen) [inline src/classes/user-config-utils.ts]:
```suggestion
  static async writeUserConfig(data : string): Promise<void> {
```

**Review by jarinykanen — CHANGES_REQUESTED** [review]

**Review by veikkahonkanen — COMMENTED** [review]
Apparently you pulled 79 to 55.

**Review by veikkahonkanen — APPROVED** [review]

File: src/classes/user-config-utils.ts [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+const userConfigPath = `.${path.sep}user-config.json`;
+const userConfigTemplate = `.${path.sep}resources${path.sep}user-config-template.json`;
+
+/**
+ * Offers functions to read and write the cli's user config
+ */
+export class UserConfigUtils {
+
+  /**
+   * runs a check after which it reads and returns the contents of user-config.json
```
Comment (jarinykanen) [inline src/classes/user-config-utils.ts]:
```suggestion
   * Runs a check after which it reads and returns the contents of user-config.json
```

File: src/classes/user-config-utils.ts [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+
+      const data = fs.readFileSync(userConfigPath, "utf8");
+      return await JSON.parse(data.toString());
+    } catch(err) {
+      throw new Error ("Error while reading user-config: " + err);
+    }
+  }
+
+  /**
+   * takes the edited version of the contents of user-config.json and writes them after a check
```
Comment (jarinykanen) [inline src/classes/user-config-utils.ts]:
```suggestion
   * Takes the edited version of the contents of user-config.json and writes them after a check
```

File: src/classes/user-config-utils.ts [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+   * 
+   * @param data should receive the edited version of user-config.json 
+   */
+  static async writeUserConfig(data : string): Promise<void> {
```
Comment (jarinykanen) [inline src/classes/user-config-utils.ts]:
```suggestion
  static async writeUserConfig(data: string): Promise<void> {
```

File: src/classes/user-config-utils.ts [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+      await this.checkUserConfig();
+
+      fs.writeFileSync(userConfigPath, data, "utf8");
+    } catch (err) {
+      throw new Error("Error when attempting to write to user-config:" + err);
+    }
+  }
+
+  /**
+   * checks that user-config.json exists and if it doesn't, creates it
```
Comment (jarinykanen) [inline src/classes/user-config-utils.ts]:
```suggestion
   * Checks that user-config.json exists and if it doesn't, creates it
```

File: src/classes/user-config-utils.ts [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+   */
+  private static async checkUserConfig() : Promise<void> {
```
Comment (jarinykanen) [inline src/classes/user-config-utils.ts]:
```suggestion
  private static async checkUserConfig(): Promise<void> {
```

File: test/user-config-utils.spec.ts [hunk 22]
```diff
@@ -0,0 +1,19 @@
+import { UserConfigUtils } from "../src/classes/user-config-utils";
+import { UserConfigJson } from "../src/interfaces/types";
+
+it("reads user-config.json", async () => {
+  const data : UserConfigJson = await UserConfigUtils.readUserConfig();
```
Comment (jarinykanen) [inline test/user-config-utils.spec.ts]:
```suggestion
  const data: UserConfigJson = await UserConfigUtils.readUserConfig();
```

**Review by jarinykanen — APPROVED** [review]

**Pull request merged**